certbot-apache/certbot_apache/interfaces.py

```diff
@@ -135,6 +125,28 @@ def name(self):
 
         raise NotImplementedError
 
+    @property
+    @abc.abstractmethod
+    def parameters(self):  # pragma: no cover
+        """
+        This property contains a tuple of parameters of this ParserNode object
+        excluding whitespaces.
+
+        :returns: A tuple of parameters for this node
+        :rtype: tuple
+        """
+
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def set_parameters(self, parameters):
+        """
+        Sets the sequence of parameters for this ParserNode object without
+        whitespaces, and marks this object dirty.
```

Comment:
I think we should describe in the docstrings of the `set_parameters` functions here or in the Google doc about what whitespace is preserved when this method is called. I think we decided that whitespace preceding the directive name is unaffected.

Reply:
The way I see it, this information belongs to the docstring of `ParserNode` itself. I'm imagining the whitespacing and metadata of the directives, blocks and comments to exist in the domain of the "main object" instead of its parameters.

I'll add it to the `ParserNode` docstring for now, and we can move it later if you disagree with this.

Reply:
What you added to the `ParserNode` docstring seems fine. (Although while you don't have to change it, it's a little general for my liking. I thought we agreed we weren't going to worry about trying to generalize this interface for more than Apache right now?)

I personally would still like a sentence saying explicitly that `set_parameters` doesn't affect the whitespace preceding the directive itself though. I think that makes most sense to include here because we're talking about the behavior of this function, but it doesn't matter to me as long as it is documented in the file.